#**Introducción a Data Science**


##12 de agosto de 2021

###Entrega Desafío V - **Francisca Pinto González**


In [ ]:
import pandas as pd
import numpy as np

In [ ]:
df = pd.read_csv("/content/worldcup2014.csv")
df

In [49]:
def media(x):
  if isinstance(x, pd.Series) is True:
    if x.dtype != "object":
      x = x.dropna()
      df_mean = round(np.mean(x), 2)
    else:
      raise ValueError("El objeto no es de tipo float o integer, selecciona otro.")
  else:
    raise ValueError("Los datos ingresados no son un objeto pd.Series.")
  return df_mean

print("La media de goles a favor es:", media(df["goles_favor"]))
print("La media de goles en contra es:", media(df["goles_contra"]))
print("La media de puntos es:", media(df["puntos"]))

La media de goles a favor es: 4.25
La media de goles en contra es: 4.25
La media de puntos es: 4.22


In [50]:
def varianza(x):
  if isinstance(x, pd.Series) is True:
    if x.dtype != "object":
      x = (x - media(x))**2
      var = np.mean(x.dropna())
    else:
      raise ValueError("El objeto no es de tipo float o integer, selecciona otro.")
  else:
    raise ValueError("Los datos ingresados no son un objeto pd.Series.")
  return var

print("La varianza en la variable goles a favor es:", varianza(df["goles_favor"]))
print("La varianza en la variable goles en contra es:", varianza(df["goles_contra"]))
print("La varianza en la variable puntos es:", varianza(df["puntos"]))

La varianza en la variable goles a favor es: 5.1875
La varianza en la variable goles en contra es: 4.8125
La varianza en la variable puntos es: 8.045899999999998


**Desafío 2**

Para la media:

In [ ]:
mean_fav_agg = df.groupby("continent")["goles_favor"].agg(lambda x: x.astype(float).mean())
mean_con_agg = df.groupby("continent")["goles_contra"].agg(lambda x: x.astype(float).mean())
mean_pts_agg = df.groupby("continent")["puntos"].agg(lambda x: x.astype(float).mean())
print("Agrupación según goles a favor: \n")
print(mean_fav_agg)
print("\n\nAgrupación según goles en contra: \n")
print(mean_con_agg)
print("\n\nAgrupación según puntos: \n")
print(mean_pts_agg)

Respecto a los resultados anteriores, se observa que:

1. El continente con mayor cantidad de goles a favor es Sudamérica.
2. El continente con mayor cantidad de goles en contra es Asia.
3. El continente con una mayor cantidad de puntos promedio es Sudamérica.



Para la desviación estándar:

In [ ]:
std_fav_agg = df.groupby("continent")["goles_favor"].agg(lambda x: x.astype(float).std())
std_con_agg = df.groupby("continent")["goles_contra"].agg(lambda x: x.astype(float).std())
std_pts_agg = df.groupby("continent")["puntos"].agg(lambda x: x.astype(float).std())
print("Agrupación según goles a favor: \n")
print(std_fav_agg)
print("\n\nAgrupación según goles en contra: \n")
print(std_con_agg)
print("\n\nAgrupación según puntos: \n")
print(std_pts_agg)

Respecto a los resultados anteriores:

1. El continente con una mayor oscilación de los resultados, respecto a los goles a favor, es Europa, por lo que la cantidad de goles a favor fue más dispar que en el resto de continentes.
2. El continente con una mayor oscilación de los resultados, respecto a los goles en contra, es Norteamérica, por lo que la cantidad de goles en contra fue dispar entre un partido y otro de los países de este continente.
3. La mayor desviación estándar en relación a la cantidad de puntos la tiene Norteamérica, por lo que en general, los equipos de este continente tuvieron resultados más dispares en los partidos que jugaron.

Para la varianza:

In [ ]:
var_fav_agg = df.groupby("continent")["goles_favor"].agg(lambda x: x.astype(float).var())
var_con_agg = df.groupby("continent")["goles_contra"].agg(lambda x: x.astype(float).var())
var_pts_agg = df.groupby("continent")["puntos"].agg(lambda x: x.astype(float).var())
print("Agrupación según goles a favor: \n")
print(var_fav_agg)
print("\n\nAgrupación según goles en contra: \n")
print(var_con_agg)
print("\n\nAgrupación según puntos: \n")
print(var_pts_agg)

**Desafío 3**

In [ ]:
pet_array = np.array(["perro", "gato"])
def generate_pet(size):
  
  x = np.array(["perro", "gato"])
  return np.random.choice(x, size)

generate_pet(20)

array(['gato', 'perro', 'gato', 'perro', 'perro', 'gato', 'perro',
       'perro', 'gato', 'perro', 'gato', 'gato', 'gato', 'gato', 'gato',
       'perro', 'gato', 'perro', 'perro', 'gato'], dtype='<U5')

In [54]:
pet_array = np.array(["perro", "gato"])
def generate_pet(size):
  np.random.seed(2)
  x = np.array(["perro", "gato"])
  return np.random.choice(x, size)

generate_pet(20)

array(['perro', 'gato', 'gato', 'perro', 'perro', 'gato', 'perro', 'gato',
       'perro', 'gato', 'perro', 'gato', 'gato', 'gato', 'gato', 'gato',
       'gato', 'gato', 'perro', 'perro'], dtype='<U5')

Respecto a las preguntas planteadas:

1. La probabilidad de elegir un perro (o un gato) al azar es de 0.5, cuando la función no tiene un parámetro "p" definido distinto a ése. Por otro lado, si al método np.random.choice no se le incorpora el parámetro "p", éste asume una probabilidad igual para todos los elementos del array, que en este caso son solo 2.
2. Al incorporar np.random.seed la función no cambia el array generado al ser ejecutada varias veces.

**Desafío 4**



In [55]:
def simulate_pets_prob(size):
  np.random.seed(1)
  array2 = generate_pet(size = ((size + 1) * 2))
  array1 = ["young_pet", "old_pet"]
  pets = [] #lista con mascotas se genera vacía al iniciar

  for i in range (1, size + 1):
    #se agregarán a la lista pets cada mascota, se generan los contadores de probabilidad, y además se generan diccionarios para hacer los condicionales de cada contador
    pets.append(array1[0] + "_" + array2[i] + ", " + array1[1] + "_" + array2[i*2])
    prob_opp = 0
    prob_ypyop = 0
    prob_ypoop = 0
    dict1 = {"young_pet_perro, old_pet_gato", "young_pet_gato, old_pet_perro", "young_pet_perro, old_pet_perro"}
    dict2 = {"young_pet_perro, old_pet_perro", "young_pet_gato, old_pet_perro"}
    dict3 = {"young_pet_perro, old_pet_perro"}

    for j in pets:

      if j in dict1:
        prob_opp += 1    
    
      if j in dict2:
        prob_ypyop += 1
    
      if j in dict3:
        prob_ypoop += 1

      array3 = np.array([prob_opp, prob_ypyop, prob_ypoop])
      array3 = array3 / len(pets)

  return array3

simulate_pets_prob(1000)

array([0.743, 0.492, 0.252])

Respecto a las consultas realizadas:

> La probabilidad de que alguna de las mascotas sea un perro es de 0.743 (o 74.3%). Por otro lado, la probabilidad de que old_pet sea un perro es de 0.492 (o 49.2%). Finalmente, la probabilidad de que ambas mascotas sean perros, es de 0.252 (o un 25.2%). A medida que crezca size, las probabilidades tenderán a 0.75, 0.5 y 0.25 respectivamente.

Respecto a las preguntas finales:

1. El escenario menos probable es el tercero (young_pet_perro y old_pet_perro simultáneamente)
2. El escenario más probable es el primero (que una o más mascotas sean perro)
3. El motivo es que en el segundo escenario existen más casos que cumplen con la condición: 3 de los 4 posibles (combinando young_pet y old_pet con perro). En cambio, en el primer escenario, solo son casos favorables la combinación young_pet_perro y old_pet_perro.
